scripts/generate-renode-scripts.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
+
+import csv
+import zlib
+import argparse
+
+def generate_ethmac(data):
+    base = get_entry_value(data, 'csr_base', 'ethmac')
+    buffer_base = get_entry_value(data, 'memory_region', 'ethmac')
+    buffer_size = get_entry_value(data, 'memory_region', 'ethmac', 1)
+    interrupt = get_entry_value(data, 'constant', 'ethmac_interrupt')
+
+    result = [
+        'ethmac: Network.LiteX_Ethernet @ {',
+        '       sysbus <{}, +0x100>;'.format(base),
+        '       sysbus new Bus.BusMultiRegistration {{ address: {}; size: {}; region: "buffer" }}'.format(buffer_base, buffer_size),
+        '    }'
+    ]
+
+    if interrupt is not None:
+        result.append('    -> cpu@{}'.format(interrupt))
+    return result
+
+def generate_peripheral(data, name, model, properties=None):
+    base = get_entry_value(data, 'csr_base', name)
+    interrupt = get_entry_value(data, 'constant', '{}_interrupt'.format(name))
+
+    result = ['{}: {} @ sysbus {}'.format(name, model, base)]
+    if properties is not None:
+        for pname, pvalue in properties.items():
+            if pvalue is not None:
+                result.append('    {}: {}'.format(pname, pvalue))
+    if interrupt is not None:
+        result.append('    -> cpu@{}'.format(interrupt))
+    return result
+
+def generate_memory_region(data, name):
+    base = get_entry_value(data, 'memory_region', name)
+    size = get_entry_value(data, 'memory_region', name, 1)
+
+    return [
+        '{}: Memory.MappedMemory @ sysbus {}'.format(name, base),
+        '    size: {}'.format(size)
+    ]
+
+def generate_silencer(data, name):
+    base = get_entry_value(data, 'csr_base', name)
+
+    return [
+        'sysbus:',
+        '    init add:',
+        '        SilenceRange <{} 0x200> # {}'.format(base, name)
+    ]
+
+def generate_flash_boot(data):
+    # TODO: this value is hardcoded, as I couldn't find it in the csv
+    base = 0x20228000
+    return [
+        'boot_flash: Memory.ArrayMemory @ sysbus {}'.format(base),
+        '    size: 0x00100000'
+        ]
+
+def generate_cpu(data):
+    kind = get_entry_value(data, 'constant', 'config_cpu_type')
+
+    if kind == 'VEXRISCV':
+        return ['cpu: CPU.VexRiscv @ sysbus']
+    elif kind == 'PICORV32':
+        return [
+            'cpu: CPU.PicoRV32 @ sysbus',
+            '    cpuType: "rv32imc"'
+        ]
+    else:
+        raise Exception('Unsupported cpu type: {}'.format(kind))
+
+def get_entry_value(data, kind, name, idx=0):
+    for entry in data:
+        if entry[0] == kind and entry[1] == name:
+            return entry[idx + 2]
+    return None
+
+def generate_repl(data):
+    return \
```

Comment:
You want code that does something like this;

```python
#! /usr/bin/env python3

data = """
csr_base,cas,0xe0006800,,
csr_base,ctrl,0xe0000000,,
csr_base,ddrphy,0xe0005800,,
csr_base,ethmac,0xe0007800,,
csr_base,ethphy,0xe0007000,,
csr_base,info,0xe0006000,,
csr_base,sdram,0xe0004000,,
csr_base,spiflash,0xe0005000,,
csr_base,timer0,0xe0002800,,
csr_base,uart,0xe0001800,,
csr_base,uart_phy,0xe0001000,,
csr_base,identifier_mem,0xe0002000,,
csr_register,cas_leds_out,0xe0006800,1,rw
csr_register,cas_switches_in,0xe0006804,1,ro
csr_register,cas_buttons_ev_status,0xe0006808,1,rw
csr_register,cas_buttons_ev_pending,0xe000680c,1,rw
csr_register,cas_buttons_ev_enable,0xe0006810,1,rw
constant,remoteip1,192,,
constant,remoteip2,168,,
constant,remoteip3,100,,
constant,remoteip4,100,,
constant,cas_leds_count,4,,
constant,cas_switches_count,4,,
constant,cas_buttons_count,4,,
constant,ethmac_rx_slots,2,,
constant,ethmac_tx_slots,2,,
"""

# Build a list of peripherals and collect all CSR registers + CSR constants associated with the peripherals
global_constants = {}
peripherals = {}
for line in csv.reader(data):
    if line[0] == "csr_base":
        _, peripheral_name, peripheral_register_base = line
        assert peripheral_name not in peripherals
        registers = {}
        constants = {}
        peripherals[peripheral_name] = (peripheral_register_base, registers, constants)

    elif line[0] == "csr_register":
        _, register_name, register_address, register_width, register_mode = line
        for peripheral_name in peripherals.keys():
            if not register_name.startswith(peripheral_name):
                continue
            peripheral_register_base, peripheral_registers, _ = peripherals[peripheral_name]

            # Strip the peripheral prefix
            register_short_name = constant_name[len(peripheral_name)+1:]

            register_offset = register_address - peripheral_register_base
            assert register_short_name not in peripheral_registers
            peripheral_registers[register_name] = (register_offset, register_width, register_mode)
            break
        else:
            # CSR Registers must be associated with a peripheral
            assert False, "CSR Register {} not associated with any peripheral {}".format(register_name, peripherals.keys())

    elif line[0] == "constant":
        _, constant_name, constant_value = line
        for peripheral_name in peripherals.keys():
            if not constant_name.startswith(peripheral_name):
                continue

            # Strip the peripheral prefix
            constant_short_name = constant_name[len(peripheral_name)+1:]

            peripheral_register_base, _, peripheral_constants = peripherals[peripheral_name]
            assert constant_short_name not in peripheral_constants
            peripheral_constants[constant_short_name] = constant_value

            break
        else:
            global_constants[constant_name] = constant_value

    else:
        assert False, "Unknown line '{}'".format(line)

import pprint
pprint.pprint(global_constants)
pprint.pprint(peripherals)

for peripheral_name, (peripheral_register_base, peripheral_registers, peripheral_constants) in peripherals.values():
    peripheral_handler = detect_peripheral_handler(peripheral_registers)
    if peripheral_handler is None:
        peripheral_hanlder = mock_handler(peripheral_register_base, peripheral_registers)

    peripheral_handler(config_file, peripheral_name, peripheral_register_base, peripheral_registers, peripheral_constants)
```

Reply:
I created a separate parse_csv function that does the job. I skipped parsing registers as it's not needed for Renode.